File: Codes in C++/Linked List/Check if a Linked List is Palindrome or not.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* head)
    {
        if(!head||!head->next)
            return head;
        ListNode* temp=reverse(head->next);
        head->next->next=head;
        head->next=NULL;
        return temp;
    }
    bool isPalindrome(ListNode* head) {
        ListNode* s,*f;
        s=head;
        f=head;
        if(!head||!head->next)
            return true;
        while(f->next!=NULL&&f->next->next!=NULL)
        {
            s=s->next;
            f=f->next->next;
        }
        f=reverse(s->next);
        s=head;
        while(f!=NULL)
        {   if(s->val!=f->val)return false;
            s=s->next;
            f=f->next;
        }
        return true;
    }
};

File: Codes in Python3/Linked List/Check if a Linked List is Palindrome or not.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        temp = self.reverse(head.next)
        head.next.next = head
        head.next = None
        return temp

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return True
        slow = head
        fast = head
        while fast.next is not None and fast.next.next is not None:
            slow = slow.next
            fast = fast.next.next
        fast = self.reverse(slow.next)
        slow = head
        while fast is not None:
            if slow.val != fast.val:
                return False
            slow = slow.next
            fast = fast.next
        return True
